File: python-ai/services/AIEmbeddingService.py
"""
AI Embedding Service
Handles embedding generation and vector storage
"""

import hashlib
from typing import List, Dict, Optional
import numpy as np
from config.aiConfig import EMBEDDING_CONFIG
from utils.logger import get_logger

logger = get_logger(__name__)


class AIEmbeddingService:
    """Service for generating embeddings"""

    def __init__(self):
        self.embedding_cache = {}
        self.vector_store = None
        self.embeddings_model = None
        self.initialize_embeddings()

    def initialize_embeddings(self):
        """Initialize embeddings model"""
        try:
            from sentence_transformers import SentenceTransformer

            model_name = EMBEDDING_CONFIG['model']
            logger.info(f"Loading embedding model: {model_name}")

            self.embeddings_model = SentenceTransformer(model_name)
            logger.info("Embedding model loaded successfully")
        except Exception as e:
            logger.error(f"Error initializing embeddings: {str(e)}")
            self.embeddings_model = None

    def generate_embedding(self, text: str) -> Optional[List[float]]:
        """Generate embedding for text"""
        if not text:
            return None

        # Check cache
        cache_key = self.hash_text(text)
        if cache_key in self.embedding_cache:
            return self.embedding_cache[cache_key]

        try:
            if self.embeddings_model:
                embedding = self.embeddings_model.encode(text).tolist()
            else:
                embedding = self.generate_simple_embedding(text)

            # Cache the embedding
            self.embedding_cache[cache_key] = embedding
            return embedding
        except Exception as e:
            logger.error(f"Error generating embedding: {str(e)}")
            return self.generate_simple_embedding(text)

    def generate_simple_embedding(self, text: str) -> List[float]:
        """Generate simple embedding (fallback)"""
        words = text.lower().split()
        embedding = [0.0] * EMBEDDING_CONFIG['dimension']

        for word in words:
            hash_val = 0
            for char in word:
                hash_val = ((hash_val << 5) - hash_val) + ord(char)
                hash_val = hash_val & hash_val

            index = abs(hash_val) % EMBEDDING_CONFIG['dimension']
            embedding[index] += 1.0

        # Normalize
        magnitude = sum(x * x for x in embedding) ** 0.5
        if magnitude > 0:
            embedding = [x / magnitude for x in embedding]

        return embedding

    def hash_text(self, text: str) -> str:
        """Hash text for caching"""
        return hashlib.sha256(text.encode()).hexdigest()

    def store_embeddings(self, document_id: str, chunks: List[Dict], embeddings: List[List[float]]):
        """Store embeddings in vector database"""
        try:
            logger.info(f"Storing embeddings: {document_id}, count: {len(embeddings)}")
            # This will be implemented with Chroma integration
        except Exception as e:
            logger.error(f"Error storing embeddings: {str(e)}")
            raise

    def search_similar(self, query_embedding: List[float], top_k: int = 5) -> List[Dict]:
        """Search similar embeddings"""
        try:
            logger.debug(f"Searching similar embeddings, top_k: {top_k}")
            # This will be implemented with Chroma integration
            return []
        except Exception as e:
            logger.error(f"Error searching similar embeddings: {str(e)}")
            return []

    def cosine_similarity(self, vec1: List[float], vec2: List[float]) -> float:
        """Calculate cosine similarity"""
        dot_product = sum(a * b for a, b in zip(vec1, vec2))
        magnitude1 = sum(x * x for x in vec1) ** 0.5
        magnitude2 = sum(x * x for x in vec2) ** 0.5

        if magnitude1 == 0 or magnitude2 == 0:
            return 0.0

        return dot_product / (magnitude1 * magnitude2)

    def clear_cache(self):
        """Clear embedding cache"""
        self.embedding_cache.clear()
        logger.info("Cleared embedding cache")

File: backend/ai/services/AIEmbeddingService.js
/**
 * AI Embedding Service
 * Handles embedding generation and vector storage
 */

const axios = require('axios');
const AIDocument = require('../models/AIDocument');
const AILogger = require('../utils/logger');
const AI_CONFIG = require('../config/aiConfig');

const logger = new AILogger('EmbeddingService');

class AIEmbeddingService {
  constructor() {
    this.embeddingCache = new Map();
    this.vectorStore = null;
    this.initializeVectorStore();
  }

  /**
   * Initialize vector store (Chroma)
   */
  async initializeVectorStore() {
    try {
      const chromaConfig = AI_CONFIG.vectorDB.chroma;
      const chromaUrl = `http://${chromaConfig.host}:${chromaConfig.port}`;

      // Test connection
      try {
        await axios.get(`${chromaUrl}/api/v1/heartbeat`);
        logger.info('Connected to Chroma vector database', { url: chromaUrl });
      } catch (err) {
        logger.warn('Chroma not available, using in-memory storage', { error: err.message });
        this.vectorStore = { vectors: new Map() };
        return;
      }

      this.chromaUrl = chromaUrl;
      this.chromaCollection = chromaConfig.collectionName;

      // Get or create collection
      await this.getOrCreateCollection();
    } catch (err) {
      logger.error('Error initializing vector store', { error: err.message });
      this.vectorStore = { vectors: new Map() };
    }
  }

  /**
   * Get or create Chroma collection
   */
  async getOrCreateCollection() {
    try {
      const response = await axios.post(
        `${this.chromaUrl}/api/v1/collections`,
        {
          name: this.chromaCollection,
          metadata: { 'hnsw:space': 'cosine' },
        }
      );
      logger.info('Created/retrieved Chroma collection', { collection: this.chromaCollection });
    } catch (err) {
      if (err.response?.status !== 409) { // 409 = already exists
        logger.warn('Error creating collection', { error: err.message });
      }
    }
  }

  /**
   * Generate embeddings for text
   */
  async generateEmbedding(text) {
    if (!text) {
      return null;
    }

    // Check cache
    const cacheKey = this.hashText(text);
    if (this.embeddingCache.has(cacheKey)) {
      return this.embeddingCache.get(cacheKey);
    }

    try {
      let embedding;

      if (AI_CONFIG.embeddings.provider === 'huggingface') {
        embedding = await this.generateHuggingFaceEmbedding(text);
      } else if (AI_CONFIG.embeddings.provider === 'openai') {
        embedding = await this.generateOpenAIEmbedding(text);
      } else {
        // Fallback: generate simple embedding
        embedding = this.generateSimpleEmbedding(text);
      }

      // Cache the embedding
      this.embeddingCache.set(cacheKey, embedding);
      return embedding;
    } catch (err) {
      logger.error('Error generating embedding', { error: err.message });
      return this.generateSimpleEmbedding(text);
    }
  }

  /**
   * Generate embedding using HuggingFace
   */
  async generateHuggingFaceEmbedding(text) {
    try {
      const response = await axios.post(
        'https://api-inference.huggingface.co/pipeline/feature-extraction',
        { inputs: text },
        {
          headers: {
            Authorization: `Bearer ${process.env.HUGGINGFACE_API_KEY}`,
          },
        }
      );

      return response.data;
    } catch (err) {
      logger.error('Error with HuggingFace API', { error: err.message });
      throw err;
    }
  }

  /**
   * Generate embedding using OpenAI
   */
  async generateOpenAIEmbedding(text) {
    try {
      const response = await axios.post(
        'https://api.openai.com/v1/embeddings',
        {
          input: text,
          model: AI_CONFIG.embeddings.openai.model,
        },
        {
          headers: {
            Authorization: `Bearer ${AI_CONFIG.llm.openai.apiKey}`,
          },
        }
      );

      return response.data.data[0].embedding;
    } catch (err) {
      logger.error('Error with OpenAI API', { error: err.message });
      throw err;
    }
  }

  /**
   * Generate simple embedding (fallback)
   * Creates a basic vector representation using word frequencies
   */
  generateSimpleEmbedding(text) {
    const words = text.toLowerCase().match(/\b\w+\b/g) || [];
    const embedding = new Array(384).fill(0);

    for (const word of words) {
      let hash = 0;
      for (let i = 0; i < word.length; i++) {
        hash = ((hash << 5) - hash) + word.charCodeAt(i);
        hash = hash & hash;
      }
      const index = Math.abs(hash) % 384;
      embedding[index] += 1;
    }

    // Normalize
    const magnitude = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0));
    if (magnitude > 0) {
      return embedding.map(val => val / magnitude);
    }

    return embedding;
  }

  /**
   * Hash text for caching
   */
  hashText(text) {
    const crypto = require('crypto');
    return crypto.createHash('sha256').update(text).digest('hex');
  }

  /**
   * Store embeddings in vector database
   */
  async storeEmbeddings(documentId, chunks, embeddings) {
    try {
      if (this.chromaUrl) {
        await this.storeInChroma(documentId, chunks, embeddings);
      } else {
        this.storeInMemory(documentId, chunks, embeddings);
      }

      logger.info('Stored embeddings', { documentId, embeddingCount: embeddings.length });
    } catch (err) {
      logger.error('Error storing embeddings', { documentId, error: err.message });
      throw err;
    }
  }

  /**
   * Store embeddings in Chroma
   */
  async storeInChroma(documentId, chunks, embeddings) {
    const ids = chunks.map((_, i) => `${documentId}-${i}`);
    const documents = chunks.map(chunk => chunk.content || chunk);
    const metadatas = chunks.map((chunk, i) => ({
      documentId,
      chunkIndex: i,
      projectName: chunk.projectName,
    }));

    await axios.post(
      `${this.chromaUrl}/api/v1/collections/${this.chromaCollection}/add`,
      {
        ids,
        embeddings,
        documents,
        metadatas,
      }
    );
  }

  /**
   * Store embeddings in memory
   */
  storeInMemory(documentId, chunks, embeddings) {
    if (!this.vectorStore.vectors) {
      this.vectorStore.vectors = new Map();
    }

    chunks.forEach((chunk, i) => {
      const id = `${documentId}-${i}`;
      this.vectorStore.vectors.set(id, {
        embedding: embeddings[i],
        content: chunk.content || chunk,
        documentId,
        chunkIndex: i,
      });
    });
  }

  /**
   * Search similar embeddings
   */
  async searchSimilar(queryEmbedding, topK = 5) {
    try {
      if (this.chromaUrl) {
        return await this.searchInChroma(queryEmbedding, topK);
      } else {
        return this.searchInMemory(queryEmbedding, topK);
      }
    } catch (err) {
      logger.error('Error searching similar embeddings', { error: err.message });
      return [];
    }
  }

  /**
   * Search in Chroma
   */
  async searchInChroma(queryEmbedding, topK) {
    const response = await axios.post(
      `${this.chromaUrl}/api/v1/collections/${this.chromaCollection}/query`,
      {
        query_embeddings: [queryEmbedding],
        n_results: topK,
      }
    );

    const results = [];
    if (response.data.ids && response.data.ids[0]) {
      for (let i = 0; i < response.data.ids[0].length; i++) {
        results.push({
          id: response.data.ids[0][i],
          content: response.data.documents[0][i],
          distance: response.data.distances[0][i],
          similarity: 1 - response.data.distances[0][i],
        });
      }
    }

    return results;
  }

  /**
   * Search in memory
   */
  searchInMemory(queryEmbedding, topK) {
    const results = [];

    for (const [id, vector] of this.vectorStore.vectors.entries()) {
      const similarity = this.cosineSimilarity(queryEmbedding, vector.embedding);
      results.push({
        id,
        content: vector.content,
        similarity,
      });
    }

    return results
      .sort((a, b) => b.similarity - a.similarity)
      .slice(0, topK);
  }

  /**
   * Calculate cosine similarity
   */
  cosineSimilarity(vec1, vec2) {
    let dotProduct = 0;
    let magnitude1 = 0;
    let magnitude2 = 0;

    for (let i = 0; i < vec1.length; i++) {
      dotProduct += vec1[i] * vec2[i];
      magnitude1 += vec1[i] * vec1[i];
      magnitude2 += vec2[i] * vec2[i];
    }

    magnitude1 = Math.sqrt(magnitude1);
    magnitude2 = Math.sqrt(magnitude2);

    if (magnitude1 === 0 || magnitude2 === 0) {
      return 0;
    }

    return dotProduct / (magnitude1 * magnitude2);
  }

  /**
   * Clear cache
   */
  clearCache() {
    this.embeddingCache.clear();
    logger.info('Cleared embedding cache');
  }
}

module.exports = AIEmbeddingService;
